### PR TITLE
move Pkg out of the sysimage (take 2)

### PR DESCRIPTION
**Pull request by KristofferC**
Will make Pkg latency bad until Pkg gets bumped to a version that includes https://github.com/JuliaLang/Pkg.jl/pull/3447. But that will happen quickly after this.

### CODE REVIEW

File: stdlib/REPL/src/REPL.jl [hunk 1]
```diff
@@ -1111,6 +1111,31 @@ function setup_interface(
                 edit_insert(s, '?')
             end
         end,
+        ']' => function (s::MIState,o...)
+            if isempty(s) || position(LineEdit.buffer(s)) == 0
+                pkgid = Base.PkgId(Base.UUID("44cfe95a-1eb2-52ea-b672-e2afdf69b78f"), "Pkg")
+                if Base.locate_package(pkgid) !== nothing # Only try load Pkg if we can find it
+                    Pkg = Base.require(Base.PkgId(Base.UUID("44cfe95a-1eb2-52ea-b672-e2afdf69b78f"), "Pkg"))
```
Comment (jonas-schulze) [inline stdlib/REPL/src/REPL.jl:1118]:
```suggestion
                    Pkg = Base.require(pkgid)
```

File: stdlib/REPL/src/REPL.jl [hunk 2]
```diff
@@ -1111,6 +1111,31 @@ function setup_interface(
...
         end,
+        ']' => function (s::MIState,o...)
+            if isempty(s) || position(LineEdit.buffer(s)) == 0
+                pkgid = Base.PkgId(Base.UUID("44cfe95a-1eb2-52ea-b672-e2afdf69b78f"), "Pkg")
+                if Base.locate_package(pkgid) !== nothing # Only try load Pkg if we can find it
+                    Pkg = Base.require(pkgid)
+                    # Pkg should have loaded its REPL mode by now, let's find it so we can transition to it.
+                    pkg_mode = nothing
+                    for mode in repl.interface.modes
+                        if mode isa LineEdit.Prompt && mode.complete isa Pkg.REPLMode.PkgCompletionProvider
```
Comment (vchuravy) [inline stdlib/REPL/src/REPL.jl:1122]: Can we do this without encoding the name?
It looks like for everything else we don't need to know about Pkg in Base which is great since we don't need to declare a compat bound ;)

Maybe we define a reference that Pkg sets in Base? If that reference is not set we do the load dance.

Comment (vtjnash) [reply stdlib/REPL/src/REPL.jl:1122]: We used to load Pkg automatically as part of launching the default REPL, which was removed when Pkg was added to the system image. We could likely still do that approach in Base (like InteractiveUtils, and the REPL itself)?

Comment (KristofferC) [reply stdlib/REPL/src/REPL.jl:1122]: > It looks like for everything else we don't need to know about Pkg in Base 

This is not Base.

> Maybe we define a reference that Pkg sets in Base? If that reference is not set we do the load dance.

What's the advantage of that over the current way?



> We could likely still do that approach in Base (like InteractiveUtils, and the REPL itself)?

I don't really want to load a package manager just because I load a REPL, no?

**Comment by KristofferC** [comment]
Seems like the RR CI gets stuck in Pkg tests:

```


From worker 7:	Username for 'https://github.com': Username for 'https://github.com': fatal: could not read Username for 'https://github.com': terminal prompts disabled
--
  | From worker 7:	fatal: could not read Username for 'https://github.com': terminal prompts disabled
  | From worker 7:	┌ Warning: Module Pkg with build ID fafbfcfd-be0a-a1a0-0028-7dc751824a03 is missing from the cache.
  | From worker 7:	│ This may mean Pkg [44cfe95a-1eb2-52ea-b672-e2afdf69b78f] does not support precompilation but is imported by a module that does.
  | From worker 7:	└ @ Base loading.jl:1942
  |  
  |  
  | Process failed to exit within 2700s, requesting termination (SIGTERM) of PID 383.
  |  
  |  
  | Sent SIGTERM to PID 383.
  |  
  | [395] signal (15): Terminated
  | in expression starting at none:1
  | unknown function (ip: 0x70000002)
  | unknown function (ip: 0x7fee6d93f2c3)
  | traced_raw_syscall at /workspace/srcdir/rr/src/preload/syscallbuf.c:338
  | syscall_hook at /workspace/srcdir/rr/src/preload/syscallbuf.c:3942 [inlined]
  | syscall_hook at /workspace/srcdir/rr/src/preload/syscallbuf.c:3926
  | unknown function (ip: 0x7fee6d939312)
  | unknown function (ip: 0x7fee6d93937c)
  | unknown function (ip: 0x7fee6d939398)
  | epoll_wait at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
  | uv__io_poll at /workspace/srcdir/libuv/src/unix/epoll.c:236
  | uv_run at /workspace/srcdir/libuv/src/unix/core.c:400
  | ijl_task_get_next at /cache/build/default-amdci4-3/julialang/julia-master/src/partr.c:469
  | From worker 7:
  | From worker 7:	[554] signal (15): Terminated
  | From worker 7:	in expression starting at /cache/build/default-amdci4-2/julialang/julia-master/julia-4fd10b97fd/share/julia/stdlib/v1.11/Pkg/test/binaryplatforms.jl:13
```

File: stdlib/REPL/src/REPL.jl (hunk 2 above)
Comment (vchuravy) [reply stdlib/REPL/src/REPL.jl:1122]: > What's the advantage of that over the current way?

That REPL doesn't need to have a written reference to an internal implementation detail of Pkg. You would be more robust if someone renamed the struct.

Comment (KristofferC) [reply stdlib/REPL/src/REPL.jl:1122]: So you want a function in Pkg

```
REPLMode_PkgCompletionProvider() = REPLMode.PkgCompletionProvider
```

and use

```
mode.complete isa Pkg.REPLMode_PkgCompletionProvider()
```

instead? I mean, I guess but it doesn't seem to be a very big difference in practice.

File: base/loading.jl [hunk 3]
```diff
@@ -1948,7 +1948,7 @@ function _require(pkg::PkgId, env=nothing)
             if pkg == concrete_pkg
                 @warn """Module $(pkg.name) with build ID $((UUID(concrete_build_id))) is missing from the cache.
                      This may mean $pkg does not support precompilation but is imported by a module that does."""
-                if JLOptions().incremental != 0
+                if true # JLOptions().incremental != 0
```
Comment (vchuravy) [inline base/loading.jl:1951]: Is this change still necessary?

File: test/runtests.jl [hunk 4]
```diff
@@ -22,6 +22,11 @@ else
     global running_under_rr() = false
 end
 
+# HACK
+if running_under_rr()
```
Comment (vchuravy) [inline test/runtests.jl:26]: We can undo this right?

**Review by vchuravy — APPROVED** [review]

File: base/loading.jl (hunk 3 above)
Comment (KristofferC) [reply base/loading.jl]: Nope

File: test/runtests.jl (hunk 4 above)
Comment (KristofferC) [reply test/runtests.jl]: Yes

**Comment by vtjnash** [comment]
This seems quite epic, and a good milestone for Julia

**Comment by oscardssmith** [comment]
how big of an effect does this have on sysimage size?
